### DISCUSSION TITLE
멀티플레이에서 서버에 접속(입장)할때 동기화 관련해서 질문합니다 :(

### DISCUSSION THREAD

**Discussion by kimyang-sun** [naverz/zepeto-studio-kor#1169] [5. Q&A]
안녕하세요! :(

제가 버튼을 눌러 의자에 앉는 기능을 구현하구 실시간으로 동기화되도록 구현을 했습니다.
멀티플레이 테스트를 할때 잘 적용이 되는 상태인데 문제는 제가 이미 앉아있는 상태에서 다른 누군가가 접속했을때
그 위치에만 제가 위치하고 앉는 제스쳐나 현재 앉아있는 상태를 표시하는 boolean 타입의 변수 등이 적용되지 않는 상태입니다.

schema에 Room State에 추가를 해주고 접속할때 그 데이터를 적용시켜주는 방식 같은데 개념 흐름이 잘 안잡히네요..
어느 부분에서 Room State에 추가를 해주고 어느 부분에서 데이터를 적용시켜줘야 될까요 ㅠ.ㅠ

우선제 코드에서 ClientStarter.ts 에서 this.room.send 를 해주고
![image](https://user-images.githubusercontent.com/56675004/184315682-434f257b-b205-4fc0-8b72-4f043d15ba8a.png)

그리고 index.ts 에서 브로드캐스트를 해주고
![image](https://user-images.githubusercontent.com/56675004/184316774-5d1cab75-4f4a-4839-b8f5-6041501f4bf6.png)

다시 ClinetStarter.ts에서 
![image](https://user-images.githubusercontent.com/56675004/184315789-1a5c3ca3-923e-4bff-a349-a0d834b4fcd9.png)

이런식으로 작업해줬습니다.

변수나 제스쳐들은 사용하는곳에서 
![image](https://user-images.githubusercontent.com/56675004/184316029-3ae68ac1-76a8-42b8-a35f-1708677a4076.png)

이렇게 this.gameObject.name 으로 오브젝트의 이름을 불러와서 위에서처럼 ClientStarter 에서 Find().GetComponent<>로 가져와서
사용해줬습니다.

**Comment by smlee2952** [comment]
제가 보기에는 앉을 때 서버에 메시지를 보내 broadcast를 하는 것 같은데, 이렇게하면 앉을 때 접속한 사람들만 메세지를 받아 동기화가 될 것 같습니다.

제가 생각하기에 가장 쉬운 방법은 OnJoinPlayer 시에 자신이 앉아 있는 상태라면 (일어서있는건 기본상태이니 아무것도 하지 않아도 괜찮을 것 같네요) SendOnSit 함수를 호출해 새로 들어온 사람에게만 메세지를 보내면 동기화 될 것 같네요.

**Reply by kimyang-sun** [reply]
@smlee2952 답변 감사합니다!
말씀해주신대로 OnJoinPlayer 에 앉아있는 상태인지를 체크하여 SendOnSit 함수를 호출해줬습니다!
위의 코드는 그대로인 상태고 ClientStarter.ts 에 OnJoinPlayer 안에 

![image](https://user-images.githubusercontent.com/56675004/184801317-56da840c-430e-4038-b4c2-b28c0ad38eac.png)

각각 앉기 관련 스크립트가 들어간 의자들에 Sit 이라는 태그로 맞춰주고 반복문으로 돌면서 isSitting 상태인 의자를
SendOnSit 을 이용해 동기화를 시켜주려고 했습니다.

그러면 첫번째로 들어온 플레이어가 의자에 앉아있고 두번째 플레이어가 접속하면 
이미 앉아있던 첫번째 플레이어에게

![image](https://user-images.githubusercontent.com/56675004/184801631-933ad014-f77e-486c-9e79-48920410b18e.png)

해당 콘솔이 잘 뜨고 SendOnSit 이 실행됩니다. (isSitting 변수가 현재 의자에 앉아있는지 여부 판단하는 변수입니다.)

하지만 이때 두번째로 접속한 플레이어에게 아래처럼 오류가 나게됩니다 :(

![image](https://user-images.githubusercontent.com/56675004/184806101-b2c68964-cb43-44e8-99ff-7731736d481b.png)

제가 생각했을때에는 현재 앉아있는 유저에게는 isSitting이 true인 의자가 있기 때문에 SendOnSit이 실행이 돼서
동기화가 되어야 한다고 생각하는데 오류가 나네요...

※ 혹시 두번째로 접속한 플레이어의 입장에서 첫번째 캐릭터가 소환이 안됐는데
 해당 캐릭터에 대해 동기화를 하려고 해서 그런거일까요???
 만약 그런거라면 캐릭터를 CreatePlayerWithUserId로 소환한 후에 정확하게 해당 코드를 실행시킬 수 있는 방법이 있을까요??

**Reply by smlee2952** [reply]
에러 로그를 보니 캐릭터를 만들기전에 호출해서 문제가 발생한 것 같습니다.

ClientStater.ts의 OnStateChange 함수 안에 있는 ZepetoPlayers.instance.OnAddedPlayer.AddListener() 안에서 처리를 해주시면 어떨까요?
제가 기억하기로 CreatePlayerWithUserId() 이후 Player 추가 리스너가 호출되는 것으로 알고있습니다.

**Reply by smlee2952** [reply]
ZepetoPlayers.instance.OnAddedPlayer.AddListener() 안에서 if (!isLocal) { } 안에서 처리해주시면 될 것 같습니다.

**Reply by kimyang-sun** [reply]
@smlee2952 안녕하세요!!
말씀해주신대로 해당 코드를 ZepetoPlayers.instance.OnAddedPlayer.AddListener() 안에서 if (!isLocal) { } 안에서 처리했는데
그대로 같은 오류가 발생해서 우선은 setTimeout 으로 의도적으로 시간을 줘서 해놓았습니다 ㅠ
답변 해주셔서 감사합니다

**Reply by smlee2952** [reply]
Player에 Sit 상태 체크할 수 있는 isSitting 변수를 추가하시는 방법도 있습니다.

한명의 플레이어가 앉을 때 SendOnSit() 호출하고 서버는 broadcast가 아닌 해당 player의 isSitting 값을 업데이트 하면
다른 플레이어들은 player의 값이 변경되었음을 체크(이미 구현되어있음, 다른 플레이어의 위치 정보를 받아 업데이트하는 부분)하고 이벤트가 발생할텐데 그때 이벤트를 발생시킨 해당 player의 캐릭터에게 앉는 애니메이션을 하게 하면 동기화가 됩니다.

**Reply by kimyang-sun** [reply]
@smlee2952 참고해서 구현해보도록 할게요! 감사합니다 :)
